Initialization of preconditioners

nfehn opened the pull request:
In general, initialization of preconditioners depends on parameters that change during a simulation (e.g. time step size, linearization vectors). Currently, we completely initialize preconditioners in the setup-phase of a solver. The usual sequence of function calls in the `Driver` during setup is

```C++
spatial_operator->setup();
time_integrator->setup();
spatial_operator->setup_solvers(time_integrator.get_...(), ...);
```
During many discussions, most recently with @richardschu, it turned out that this procedure is quite impractical and I wonder whether we could find a better solution.

So far, the preconditioner already receives automatically the relevant information for preconditioner updates during time-dependent simulations (time steppping) or nonlinear solvers (linearization vector). Hence, we principally already have the functionality to not trigger the preconditioner init/update in the driver, but let it be triggered somewhat "automatically" (typically realized by preconditioners having a reference/pointer to the `underlying_operator`, which is the object that has to be up-to-date and from which parameters can be queried). However, what we would need in my opinion is being able to ask the preconditioner whether it as been initialized so far, such that a Krylov solver can then trigger the initialization of the preconditioner.

I think such a procedure would simplify or clarify the code overall. One downside would be that in terms of compute timings, setup costs in the sense of preconditioner initialization would appear as "solver costs" or "time integration costs", even though the initialization could be done prior to starting the time loop (exactly the procedure we do currently). However, the idea would be that for preconditioners that do not depend on dynamic parameters (say the inverse mass matrix preconditioner), we can do the initialization of the preconditioner already in the constructor. Hence, for these simple cases, `preconditioner.is_initialized()` called in the Krylov solver during the first time step would evaluate to true and nothing would have to be done.

richardschu reviewed (commented):
as discussed, the `update_preconditioner(bool const update_preconditioner) const override` will be removed. Otherwise good to go!

kronbichler commented on include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_base.h:
```diff
@@ -51,6 +69,12 @@ class PreconditionerBase
     return std::make_shared<TimerTree>();
   }
+
+private:
+  virtual void
+  do_initialize() = 0;
```
I guess this function will eventually be `protected` rather than `private`, because we expect the derived classes to override the function?

kronbichler reviewed (commented):
I think this can make sense, albeit I would not overengineer things and flexibly see for the needs. I see no problem in solver costs being reported with preconditioner initialization/update, because those can always be annotated more clearly in case of need by changing the source code for the timing purposes (or a suitable longer-term solution).

The question not clear to me yet is what the mechanism to indicate that the preconditioner has become out of date, e.g. if the mesh (or time step) has changed, is supposed to be. We have the `update` functionality that we used before to update the parameters, but the question is whether we would expect the preconditioner to make a full setup (potentially changing vector sizes, parallel partitioning, etc.) during `update` then? Or is the preconditioner then destroyed when a full setup including mesh + dof_handler update is done?

nfehn commented on include/exadg/solvers_and_preconditioners/solvers/iterative_solvers_dealii_wrapper.h:
```diff
@@ -272,6 +277,11 @@ class SolverGMRES : public SolverBase<VectorType>
     }
     else
     {
+      if(not preconditioner.is_initialized())
+        preconditioner.initialize();
+      else if(update_preconditioner)
+        preconditioner.update();
```
aims to make sure that preconditioner is not initialized and - immediately after this - updated again.

nfehn commented:
With overengineering, do you refer to the old or to the intended new implementation? The intention of this PR would be to simplify or clarify the code overall. Overengineering is in my understanding needless complexity and should be avoided of course.

You mentioned adaptive mesh refinement: When setting `refine_mesh_every_time_steps = 10` and `update_preconditioner_every_time_steps = 5`, the code would currently do a complete initialization of preconditioners every 10th time step, and do additional updates every 5th time step. This would mean there could be time steps where the diagonal (or Chebyshev smoother setup) are computed twice per time step. I think what we all want is to be able to detect such cases and avoid unnecessary initalization/update work? The aim of this PR is to find a good solution for such problems with a design where the user can understand the code and is able to use it correctly. The if-elseif-solution used here https://github.com/exadg/exadg/pull/599/files#r1394190961 is one suggestion for this problem.

Originally, I did not intend to address in this PR the invalidation of preconditioners (in case something fundamental happens like the partitioning changes). I believe we would indeed need such a mechanism and that the design could look similarly as in the present PR (e.g. an additional member function `set_uninitialized()` setting the private member function `has_been_initialized = false`). I could already implement this here. 

I think we need mechanisms to distinguish between mandatory and optional preconditioner "updates". My current understanding is that

- `initialize()` does a complete initialization, which is necessary in the very beginning and maybe after the grid/partitioning has been changed. The preconditioner is then ready-to-use after this function.
- `update()`: updates the preconditioner after certain parameters (like the time step size) have changed, e.g. recalculate the diagonal.

nfehn commented:
> as discussed, the update_preconditioner(bool const update_preconditioner) const override will be removed. Otherwise good to go!

It turned out that this is not possible because the current Newton solver implementation needs to be able to trigger a separate preconditioner update.

nfehn commented:
The motivation for this PR is probably best seen from commits https://github.com/exadg/exadg/pull/599/commits/2d6a0cdfa2240ddd0fefe4c53d6a62eebf518226 and https://github.com/exadg/exadg/pull/599/commits/bd777f78f18f0ec57815424a65331c74bac972e8.

nfehn commented:
Commit https://github.com/exadg/exadg/pull/599/commits/6e8911bde92963468eecdc7aba0d21b7ceaa02ee fixes a (potential) bug, principally already available in the previous implementation in my opinion.

nfehn commented on include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.cpp:
```diff
@@ -213,13 +215,9 @@ OperatorCoupled<dim, Number>::rhs_stokes_problem(BlockVectorType & dst, double c
+                                                       BlockVectorType const & src) const
 {
   // (1,1) block of saddle point matrix
-  this->momentum_operator.set_time(time);
-  this->momentum_operator.set_scaling_factor_mass_operator(scaling_factor_mass);
```
In the previous implementation, `scaling_factor_mass` is only set in `apply_linearized_problem(), which is the `vmult()` function of the Krylov solver. Hence, the preconditioner does not know about this factor.

nfehn replied:
no, it is exactly the `momentum_operator` that needs it. However, when setting these parameters in the `vmult()` function, the preconditioner can not query this data from the `momentum_operator` during `preconditioner->update()`.

nfehn replied:
... because `preconditioner->update()` is called before the code runs into `operator->vmult()` in the Krylov solver.

nfehn commented on include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.cpp:
```diff
@@ -171,11 +171,13 @@ unsigned int
                                                           double const & scaling_factor_mass)
 {
-  // Update linear operator
-  linear_operator.update(time, scaling_factor_mass);
```
This line did actually not call `momentum_operator->set_scaling_factor_mass_operator()` in the previous implementation, such that the following line `linear_solver->update_preconditioner()` used a potentially wrong `scaling_factor_mass`. This can have a significant influence on iteration counts.

nfehn commented:
I think I addressed everything and the code would be ready for review.

kronbichler reviewed (approved):
Looks good to me, this is a good simplification.

nfehn commented:
I realized that for some reasons, certain preconditioning options have so far not been supported for certain modules. I added these options in commit https://github.com/exadg/exadg/pull/599/commits/ddf0704e93a844cd385596224cd2a4adcd5f55e6.

@bergbauer For the Structure module, I added `AdditiveSchwarz`.

richardschu commented on include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.cpp:
```diff
@@ -258,8 +250,9 @@ OperatorCoupled<dim, Number>::solve_nonlinear_problem(BlockVectorType &  dst,
   // Update nonlinear operator
   nonlinear_operator.update(rhs_vector, time, scaling_factor_mass);
 
-  // Update linear operator
```
I do not get line 262 in the old code - is `linear_operator` the *linearized* operator? Why was the `momentum_operator` not updatedpreviously? Is that the bug you were mentioning before?

nfehn replied:
The linear operator is just needed as the class handed over to the Krylov solver as the matrix or operator, providing the function `vmult()`. For our matrix-free implementation, `momentum_operator` is the relevant spatial operator.

The problem or potential bug was that `linear_operator->update()` did not immediately update the `momentum_operator`, which however would have been necessary for a correct update of the preconditioner.

richardschu commented on include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:
```diff
@@ -1577,6 +1568,17 @@ SpatialOperatorBase<dim, Number>::setup_projection_solver()
                                        projection_operator->get_dof_index(),
                                        projection_operator->get_quad_index());
     }
+    else if(param.preconditioner_projection == PreconditionerProjection::PointJacobi)
```
Why is this showing up here? You just added a new elementwise preconditioner since you were at it?

nfehn replied:
yes. I also commented in this commit here https://github.com/exadg/exadg/pull/599#issuecomment-1821040787

richardschu commented on include/exadg/operators/inverse_mass_operator.h:
```diff
@@ -277,7 +277,8 @@ class InverseMassOperatorHdiv
       preconditioner =
         std::make_shared<JacobiPreconditioner<MassOperator<dim, n_components, Number>>>(
-          mass_operator);
+          mass_operator, true);
```
'/* initialize_preconditioners */' ?

nfehn replied:
I am somewhat undecided. I  can change it if you want. If so, we should do it consistently (i.e. there are probably also other places).

richardschu replied:
we can do it or not do it, I do not think this is too strict ... but in line 139 above you had it and here in line 280 its not. So decide for one variant and make it consistent in the file :smile:

richardschu commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -82,13 +83,13 @@ MultigridPreconditionerBase<dim, Number, MultigridNumber>::initialize(
 
-  this->initialize_transfer_operators();
+  this->initialize_coarse_solver(operator_is_singular, initialize_preconditioners);
 
```
Why is the sequence here changed as well? Does that have a reason?

nfehn replied:
Not strictly necessary, but I wanted to clarify the code by first doing the initialization steps not depending on `initialize_preconditioners`, and then do the rest that depends on `initialize_preconditioners`.

richardschu commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -662,6 +665,10 @@ void
+  AssertThrow(not this->update_needed,
+              dealii::ExcMessage(
+                "Multigrid preconditioner can not be applied because it needs to be updated."));
+
```
good assert!

richardschu commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -670,6 +677,10 @@ unsigned int
+  AssertThrow(not this->update_needed,
+              dealii::ExcMessage(
+                "Multigrid preconditioner can not be applied because it needs to be updated."));
+
```
good assert!

richardschu commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -806,21 +820,20 @@ MultigridPreconditionerBase<dim, Number, MultigridNumber>::initialize_coarse_sol
-      }
-      else if(data.coarse_problem.amg_data.amg_type == AMGType::BoomerAMG)
+      if(data.coarse_problem.amg_data.amg_type == AMGType::ML or
+         data.coarse_problem.amg_data.amg_type == AMGType::BoomerAMG)
```
:+1:

richardschu commented on include/exadg/solvers_and_preconditioners/preconditioners/additive_schwarz_preconditioner.h:
```diff
@@ -34,32 +34,41 @@ class AdditiveSchwarzPreconditioner : public PreconditionerBase<typename Operato
+      not this->update_needed,
+      dealii::ExcMessage(
+        "Additive Schwarz preconditioner can not be applied because it needs to be updated."));
+
```
:+1:  better naming and good assert!

richardschu commented on include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h:
```diff
@@ -114,23 +113,25 @@ class PreconditionerML : public PreconditionerBase<Number>
   {
-    // clear content of matrix since the next calculate_system_matrix-commands add their result
+    // clear content of matrix since calculate_system_matrix() adds the result
     system_matrix *= 0.0;
```
is this cheaper than `=0` or equivalent?

richardschu commented on include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h:
```diff
@@ -220,6 +213,21 @@ class PreconditionerBoomerAMG : public PreconditionerBase<Number>
+    // skip the processes that do not participate in the matrix and have size
+    // zero
+    if(system_matrix.m() > 0)
+      system_matrix = 0.0;
```
why not `*= 0` as above?

nfehn replied:
I would suggest to look at this somewhere else, or did I change the code here in this PR?

richardschu replied:
no, code was like this before - I was just wondering.
We could open up an issue and see what the others think?

richardschu commented on include/exadg/structure/spatial_discretization/operator.cpp:
```diff
@@ -256,13 +256,15 @@ Operator<dim, Number>::setup_operators()
 
     mass_operator.set_scaling_factor(param.density);
 
+    // TODO: this comment can be removed once we merge setup() and setup_solvers()
```
`setup_solver()`

nfehn replied:
good hint!

nfehn replied:
please check again commit https://github.com/exadg/exadg/pull/599/commits/0b89496c22a68c1db3c494c66299aa90465d7a86

richardschu commented on include/exadg/structure/spatial_discretization/operator.cpp:
```diff
@@ -373,12 +358,26 @@ Operator<dim, Number>::initialize_preconditioner()
+  }
+  else if(param.preconditioner == Preconditioner::AdditiveSchwarz)
+  {
+    if(param.large_deformation)
```
:+1:  cool to see AdditiveSchwarz added as well!

richardschu commented on include/exadg/incompressible_navier_stokes/user_interface/enum_types.h:
```diff
@@ -370,6 +370,7 @@ enum class PreconditionerPressurePoisson
 {
   None,
   PointJacobi,
+  BlockJacobi,
```
was this just forgottten?

nfehn replied:
no, I added this preconditioner variant here. You are right in the sense that this type of preconditioner would have been available before the present PR. The reason is probably historically, i.e. the block-Jacobi preconditioner was added later to ExaDG, while for the pressure Poisson equation, the geometric multigrid preconditioner is our working horse.

richardschu commented on include/exadg/solvers_and_preconditioners/solvers/iterative_solvers_dealii_wrapper.h:
```diff
@@ -122,9 +122,12 @@ class SolverCG : public SolverBase<VectorType>
+    if(solver_data.use_preconditioner)
     {
-      preconditioner.update();
+      if(preconditioner.needs_update() or update_preconditioner)
```
ok so actually, `needs_update` is only `false` initially and causes a preconditioner update in the very first solver call. Afterwards, we force the update from the linear solver call, right?

nfehn replied:
Yes, exactly. Let's put it this way: "The user is the king.", i.e. if the user says the preconditioner should be updated, we will do it. "With great power comes great responsibility", though ;)

richardschu reviewed (approved):
Please clearify some questions I added, otherwise good! I am happy that we could remove the complications in the drivers. Very cool!

richardschu reviewed (approved):
thanks for addressing my questions, check out the comments: one regarding consistency of `false /* initialize_precodnitioner */` and one regarding the potential issue on `matrix = 0;` or `matrix *= 0;`
This PR is ready to merge :100:
